### ISSUE TITLE
Error when trying to run qsub for ED2 XML

### ISSUE THREAD

**Issue by Viskari** [PecanProject/pecan#502]
Currently, when I set in my XML for pecan to run ED2 with a qsub, pecan fails to submit the run with the following error message:

[1] "----------------------------------------------------------------------"
[1] "STARTING MODEL"
[1] "----------------------------------------------------------------------"
[1] "-------------------------------------------------------------------"
[1] " Starting model runs ED2"
[1] "-------------------------------------------------------------------"
  |======================================================================| 100%[1] "ENS-00001"
sh: : command not found
Error in system2(cmd, c(qsub, file.path(settings$rundir, format(run, scientific = FALSE),  :
  error in running command
Calls: start.model.runs -> system2
[1] "----------------------------------------------------------------------"
[1] "DONE"
[1] "----------------------------------------------------------------------"

The confounding issue is that our system does recognize sh and if I go the actual run folder and run the job script there, it does execute it, although far too fast.

Thought to ask if there are any thoughts on what might be the issue at hand?

**Comment by mdietze** [comment]
What does the logfile.txt say? Is your system still firewalled or are the outputs now accessible through the PEcAn web interface?

**Comment by Viskari** [comment]
It doesn't produce a logfile and the firewall is still an issue, so the
outputs can't be accessed through the webinterface.

I can now add that we can run the ensemble sequantally, so it isn't an
ensemble issue.

**Comment by serbinsh** [comment]
Mike

We are running through the shell, not the web interface.

S

[iPhone]

On Jun 18, 2015, at 2:02 PM, Viskari <notifications@github.com<mailto:notifications@github.com>> wrote:

It doesn't produce a logfile and the firewall is still an issue, so the
outputs can't be accessed through the webinterface.

I can now add that we can run the ensemble sequantally, so it isn't an
ensemble issue.

**Comment by mdietze** [comment]
@Viskari are you *sure* there's no logfile? that should be in the outdir not rundir and opening that file is the very first thing that job.sh does. If you're not getting that file when you run job.sh on the command line then you need to open up that script and check what's going on.  Do you not have bash installed? Is it not at /bin/bash?

@serbinsh it doesn't matter whether you run PEcAn through the web interface or workflow.R, the results still show up in the web interface  ;-)

**Comment by Viskari** [comment]
So I tried it again and now it crashed at a different part of the code, but creates logfile. The error message now is

[1] "----- Variable: NPP"
[1] "----- Running ensemble analysis for site:  Willow Creek (US-WCr)"
Error in hist.default(unlist(ensemble.output), xlab = units, main = "",  :
  'x' must be numeric
Calls: run.ensemble.analysis -> hist -> hist.default

If I don't have it run with qsub, it runs perfectly fine. I have no idea why the error changed as I used the exactly same addition to the XML file, but such is life.

**Comment by Viskari** [comment]
And forgot to add, the logfile shows it running for a month and half, but doesn't contain any error messages.

**Comment by Viskari** [comment]
@robkooper @mdietze @serbinsh 

I've been going through the pecan code trying to figure out this error and I think I might have gotten somewhere.

So, when I look at the run.ensemble.analysis, I had difficulties nailing it completely down because I can't find where it gives the value to ensemble.output in the file, but by comparing qsub runs with direct runs it seems that currently pecan runs the ensemble.analysis almost immediately after the qsub run has been submitted without waiting for the run the actually be finished.

Could this be an issue with a switch or something related to how our system is setup?

**Comment by mdietze** [comment]
@Viskari looking in start.model.runs, if qsub is used then the internal variable jobids should be set, the system should display the runs in the logfile under "Waiting for the following jobs:", and then it should proceed to check on all the jobs until they're finished, and only then proceed on.  I'd first check that PEcAn is finding your job id's at all (if not, you probably need to mess with the settings under settings$run$host$qsub). If it's generating the list but not checking their status correctly you'll want to mess with settings$run$host$qstat. The variables in the settings list map directly to the equivalent nested tags in the pecan xml.

**Comment by robkooper** [comment]
One other thing to check, are you specifying the qsub command in your pecan.xml file? You will need to specify at least an empty qsub command. If you specify the actual command with parameters to use, make sure you add the flag -S /bin/bash which will make sure that script is executed with the bash shell.

**Comment by Viskari** [comment]
So I checked the run and it does print out the waiting for the following
jobs with the job ids showing up there and by checking the logfiles, it is
starting the runs, with qsub it just starts checking the output before the
run is finished. A copy of the output is below

2015-06-22 11:53:13 DEBUG  [#4: start.model.runs] :
   Waiting for the following jobs: 34.test.esh.bnl.gov
2015-06-22 11:53:23 DEBUG  [#4: start.model.runs] :
   Job 34.test.esh.bnl.gov for run ENS-00001 finished
[1] "----------------------------------------------------------------------"
[1] "DONE"
[1] "----------------------------------------------------------------------"
[1] "----------------------------------------------------------------------"
[1] "STARTING OUTPUT"
[1] "----------------------------------------------------------------------"
2015-06-22 11:53:23 INFO   [read.ensemble.output] :
   reading ensemble output from run id: ENS-00001
2015-06-22 11:53:23 ERROR  [read.output] :
   no netCDF files of model output present

And here is the section in the XML file. I literally copied it from the
example contained in the wiki.



    </host>    <host>
      <name>localhost</name>
      <rundir>/data/tviskari/pecan/run/</rundir>
      <outdir>/data/tviskari/pecan/out/</outdir>
      <qsub>qsub -N @NAME@ -o @STDOUT@ -e @STDERR@</qsub>
      <qsub.jobid>lionxo_test ([0-9]+) .*</qsub.jobid>
      <qstat>qstat -j @JOBID@ 2>1 >/dev/null || echo DONE</qstat>
    </host>

**Comment by mdietze** [comment]
Can you confirm that the job.sh run by hand completes correctly? 

Can you also check that when you start the job.sh in your queue by hand whether "qstat -j @JOBID@ 2>1 >/dev/null || echo DONE" works correctly (where @JOBID@ is substituted with your job id)?  

Right now in your log it's looking like PEcAn is getting the message that the job is done, so we need to isolate whether the job is wrong or the message.  All the output processing is done WITHIN job.sh so if that script finishes correctly then PEcAn won't be accessing the output before the run is finished.

**Comment by serbinsh** [comment]
Toni

Have you checked qstat -u '*' 

to check that the jobs are being submitted?

**Comment by serbinsh** [comment]
```
<run>
    <start.date>1998-07-01 00:00:00</start.date>
    <end.date>2005-01-01 00:00:00</end.date>
    <dbfiles>/data/serbin-data/Modeling/dbfiles</dbfiles>
    <site>
      <id>676</id>
      <name>Willow Creek (US-WCr)</name>
      <lat>45.92</lat>
      <lon>-90.45</lon>
      <met>/data/source/Sites/US-WCr/ED_MET_DRIVER_HEADER</met>
      <met.start>1998</met.start>
      <met.end>2006</met.end>
    </site>
    <host>
      <name>localhost</name>
      <qsub>qsub -N @NAME@ -o @STDOUT@ -e @STDERR@</qsub>
<!--      <qsub.jobid>PEcAn ([0-9]+) .*</qsub.jobid> -->
      <qstat>qstat -j @JOBID@ 2>1 >/dev/null || echo DONE</qstat>
    </host>
  </run>
```

Here is a version that I used on a server system I setup at UW Madison in 2013.  I was using Sun GridEngine.  Note that I commented out the <qsub.jobid> line.

**Comment by robkooper** [comment]
Try using 

```
    <host>
      <name>hostname</name>
      <qsub/>
    </host>
```

it should use the defaults as specified at https://github.com/PecanProject/pecan/wiki/PEcAn-Configuration#run-setup

**Comment by Viskari** [comment]
So I've been going over the different variations in the XML based on Rob's,
Shawn's and Edgar's feedback, and in all cases it still crashes at the same
point. It still seems that it submits the run, but tries to read results
before the run is finished.

**Comment by robkooper** [comment]
What is the actual output from qsub? It could be that the pattern matcher is not correct for the jobid.

**Comment by Viskari** [comment]
Where is it visible in? I checked the logfile, but that only shows the
interrupted ED2 run.

-- Toni

**Comment by Viskari** [comment]
And a slight backtrack. I got it to run by combining the suggestions from
Rob and Edgar.

So if I run with the tag <qsub/> outside the localhost, then I am able to
run is via qsub where it waits until the run is finished before PEcAn tries
to readin the outputs.

And thanks again for all the help.

**Comment by robkooper** [comment]
I don't think it uses qsub at that point, but simply runs the job on the head node. Can you check your Rout file to make sure it actually uses qsub?

**Comment by Viskari** [comment]
You are correct, although this confuses me so much. At the moment, if I run
pecan like that, when checking with qstat I can see that there are no
submitted jobs. However, the reason I thought it was working was because
when I first got it to work and checked it then with qstat, that showed the
queued job running, which it doesn't anymore.

Where Rout file? I tried checking all the folders in the output pecan
folder, but couldn't find it.

-- Toni

**Comment by robkooper** [comment]
how do you start pecan? the output from running pecan should contain the right information. If you run it from the command line make sure you pipe the output to a log file you can later look at.

**Comment by Viskari** [comment]
I run pecan at the moment with command

./workflow.R pecan64.test.ed.xml

And the screen gives me a lot of information, but it doesn't show me the
exact command it runs qsub as. The only thing on that it gives me is

2015-06-22 17:58:05 DEBUG  [#4: start.model.runs] :
   Waiting for the following jobs: 42.test.esh.bnl.gov
2015-06-22 17:58:15 DEBUG  [#4: start.model.runs] :
   Job 42.test.esh.bnl.gov for run ENS-00001 finished

**Comment by robkooper** [comment]
good this is exactly what I need. The output from qsub is `Waiting for the following jobs: 42.test.esh.bnl.gov`, which needs to be parsed by <qsub.jobid>, the default is `Your job ([0-9]+) .*` which does not match.

Try setting it to:
```
<qsub.jobid>Waiting for the following jobs: ([0-9]+).*</qsub.jobid>
```
or maybe
```
<qsub.jobid>Waiting for the following jobs: (.*)</qsub.jobid>
```

**Comment by Viskari** [comment]
Tried both settings, still the same error.

**Comment by robkooper** [comment]
when you do a qsub what do you use to see the qstat. Right now it will use `qstat -j @JOBID@` where @JOBID@ is replaced with the id parsed from qsub.jobid. So in your case what does it do when you type `qstat -j 42.test.esh.bnl.gov`?

**Comment by Viskari** [comment]
Normally if I check job status I just use qstat. However, when I tried with
the command you gave, I get an error message that qstat doesn't recognize
the flag -j.

**Comment by robkooper** [comment]
That is the problem,the -j flag is used to pass the jobid so we can check to see if the job is finished. If the job is not there and throws an error it will return DONE to PEcAn, which is why it continues without waiting for the job. Send me by email the output from `man qsub` so I can see what the options are.

**Comment by robkooper** [comment]
what happens if you type `qstat 42.test.esh.bnl.gov` | echo DONE does it print DONE?

In you pecan.xml can you use the following for qstat:
```
qstat @JOBID@ &> /dev/null || echo DONE
```

**Comment by Viskari** [comment]
Sorry about the delay, I was having slight issues testing this as the
workflow started crashing if I added any qstat flag there despite yesterday
being able to handle them without issues. Because of course it did.

I ended up hardcoding the qstat version in to start.model.runs, but it
didn't seem to work.

**Comment by Viskari** [comment]
Okay, I did a little bit more experimenting and the command suggested by
Rob actually works on the command line where it gives empty as long as the
specified run is still in the system, but gives DONE when that run is
finished. So conceptually it should be working and as R is using the system
call, it should produce the same from there.

Next I will thus try to figure out how the pecan run differs from this.

**Comment by Viskari** [comment]
And finally progress, of sorts.

With Shawn's help I was able to figure out what was causing problems with
my XML file, I had to write 2>1 instead of &>, and I gave the following
flag in the XML:

 <qstat>qstat @JOBID@ 2>1 /dev/null || echo DONE</qstat>

And now it runs and waits for the signal from the system before moving on
the result analysis.

To avoid future issues, I would recommend removing the assumed -j flag from
pecan as when I was checking I actually found a number of qstat manuals
that don't list that flag and it seems to work fine even without it.

Thanks for all the help with the matter.
